packages/shared-base/src/utils/renderString.ts

```diff
@@ -0,0 +1,9 @@
+type ExtraKeys<S extends string> = S extends `${infer HEAD}{${infer K}}${infer TAIL}`
+    ? K extends '}'
+        ? never
+        : ExtraKeys<HEAD> | K | ExtraKeys<TAIL>
+    : never
+
+export function renderString<T extends string>(template: T, data: Record<ExtraKeys<T>, string | number>) {
```

Comment:
suggest, rename the function name to `renderTemplateString`

Reply:
What about this already existing solution [_.template](https://lodash.com/docs/4.17.15#template).

Reply:
lodash template function use eval function

Reply:
We might rewrite with i18n instead. I18n engine itself provides template rendering. And these places miss i18n indeed.